### PR TITLE
Remove list of subpackages from parent Kptfile

### PR DESCRIPTION
**Pull request by mortent**
This PR updates the handling of subpackages. It no longer keeps a list of subpackages in the parent Kptfile, but rather a subpackage is just like any other package that exists in a subdirectory of another package. Making this change required quite a few updates, so unfortunately this PR is something large. There will also be follow-up PRs to clean up some things and add more tests.

Some issues that are not covered by this PR, but will be handled in follow-up PRs:
 * This still uses string type for paths. Eventually we want to use the `UniquePath` type for this.
 * There is duplication in the `internal/util/pkgutil` package and it has overlap with the `internal/pkg` package. That will be addressed in a follow-up PR.
 * There is still separate code for handling `get` and `update`. Eventually we want to reconcile this into one.
 * More tests for handling complex update scenarios.

### CODE REVIEW

File: internal/testutil/setup_manager.go
Comment (phanimarupaka) [inline internal/testutil/setup_manager.go:217]: kptfilev1alpha2?

File: internal/util/pkgutil/pkgutil.go
Comment (phanimarupaka) [inline internal/util/pkgutil/pkgutil.go:31]: What do you mean by remote subpackages ? Isn't that concept going away?

File: internal/util/pkgutil/pkgutil.go
Comment (phanimarupaka) [inline internal/util/pkgutil/pkgutil.go:47]: What are excluded dirs ? Can you add more comments as to what this method is doing, explaining the variables.

File: internal/util/pkgutil/pkgutil.go
Comment (phanimarupaka) [inline internal/util/pkgutil/pkgutil.go:116]: Please explain what this is doing.

File: pkg/kptfile/kptfileutil/util.go
Comment (phanimarupaka) [inline pkg/kptfile/kptfileutil/util.go:208]: Why is merge not taking origin ? It should be 3-way merge correct ?

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [inline pkg/api/kptfile/v1alpha2/types.go:43]: Will Kptfile have both Upstream and UpsteamLock ? Why can't we just go down the current(master) route and just have Upstream ?

**Review by phanimarupaka — CHANGES_REQUESTED** [review]

File: internal/util/pkgutil/pkgutil.go
Comment (mortent) [reply internal/util/pkgutil/pkgutil.go:31]: We still use the distinction between local and remote packages, which is determined by whether the Kptfile has an `upstream` section defined. During updates, we do want to update subpackages, and without a list in the parent, we find those packages by scanning the filesystem.

File: internal/util/diff/diff.go
Comment (frankfarzan) [inline internal/util/diff/diff.go:307]: Probably want a WriteKptFile() method to Package as counterpart to:

https://github.com/GoogleContainerTools/kpt/blob/f652b15b033a763967152e397d05320fe53166f4/internal/pkg/pkg.go#L91

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:43]: We need both information about the upstream used for the current version of the local package, which used to be the `upstream` section of the package Kptfile, but which is now included in the `upstreamLock` section. This is information that is always automatically updated by kpt.
We also need to allow the user to declaratively specify the desired upstream fro the package. This information used to be in the `subpackages` section of the parent Kptfile, but we have now moved it into the package itself. The update process consists of two steps, first update the "desired" upstream by changing the information in the `upstream` section (either manually or through kpt porcelain commands), and then ask kpt to perform the update by pulling down the desired version and performing the merge.

File: internal/util/diff/diff_test.go
Comment (frankfarzan) [inline internal/util/diff/diff_test.go:169]: Might want to consider reducing verbosity somehow: `kptfilev1alpha2`package name is repeated a lot

File: internal/util/fetch/fetch.go
Comment (frankfarzan) [inline internal/util/fetch/fetch.go:37]: Mentioned previously: This should use the Package type.

Per the PR description, it's fine to have this in follow up PR since this is already a large change.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (phanimarupaka) [reply pkg/api/kptfile/v1alpha2/types.go:43]: So a Kptfile contains both Upstream and UpstreamLock section at the same time ? How does a Kptfile look like? I think that is confusing. How would a user know what to edit ? 

```
The update process consists of two steps, first update the "desired" upstream by changing the information in the upstream section (either manually or through kpt porcelain commands), and then ask kpt to perform the update by pulling down the desired version and performing the merge.
```

This is the case even with current form of kpt. I think having just one Upstream section is simple and intuitive. Users can directly edit Upstream section. Kpt can see that section, perform update/merge, update commit hash.

File: internal/util/pkgutil/pkgutil.go
Comment (frankfarzan) [inline internal/util/pkgutil/pkgutil.go:15]: This should be folded into Package type. Following comments about the need for better abstractions for a Package and avoiding passing raw strings around everywhere.

Also see:

https://github.com/GoogleContainerTools/kpt/blob/f652b15b033a763967152e397d05320fe53166f4/internal/pkg/pkg.go#L162

Comment (frankfarzan) [reply internal/util/pkgutil/pkgutil.go:15]: One suggestion to reduce the size of this PR: Have the changes to the Package type as as separate PR. The remaining code can use the high level Package abstraction.

**Review by frankfarzan — CHANGES_REQUESTED** [review]
I didn't look at everything in detail. Added a suggestion for narrowing the scope of this PR.

File: internal/cmdupdate/cmdupdate.go
Comment (mortent) [reply internal/cmdupdate/cmdupdate.go:91]: I considered adding this as part of the PR, but this would make an already massive PR even larger. This is mentioned in the PR description.
I do want to consolidate the code in `kpt/internal/pkg/pkg.go` with the update code, but it will take some effort. The new pkg code has added code that seems to duplicate existing functionality that needs to be cleaned up. Also, the package abstraction doesn't help when working with nested packages, which is where most of the filepath manipulation happens. In this case we do need to work with relative paths, but the `DisplayPath` abstraction doesn't really help.

I also have some concerns about how we currently handle relative and absolute paths in the `internal/pkg` code. We should translate any relative paths in the cobra commands and then only work with absolute paths. If user feedback makes more sense with relative paths, we should translate it back in the cobra commands. Kpt used to mix relative and absolute paths and it makes testing very difficult.

I guess this was just a long way of saying that I do plan on aligning with the package abstraction, but I want to do that separately. I also want to get this in first, so I can iterate on code that actually does the right thing instead of continue on working with the old design.

File: internal/util/get/get.go
Comment (droot) [inline internal/util/get/get.go:75]: Interesting... so we write a default Kptfile with upstream info, and then fetch does the rest and merged the kptfile.
This is more of a note for myself.

File: internal/util/diff/diff.go
Comment (mortent) [reply internal/util/diff/diff.go:307]: Yeah, we already have code for handling the Kptfile in `pkg/kptfile`, including `ReadFile` and `WriteFile` functions. But we seem to also have added this in `internal/pkg`.

File: internal/util/diff/diff_test.go
Comment (mortent) [reply internal/util/diff/diff_test.go:169]: Yeah, it is a bit verbose, but it kinda follows the pattern used in kubernetes. Maybe we should standardize on `kfv1alpha2` as the prefix. The default import name of `v1alpha2` is a bit unfortunate, so we should use a consistent alias.

File: internal/util/pkgutil/pkgutil.go
Comment (mortent) [reply internal/util/pkgutil/pkgutil.go:15]: yeah, the `internal/util/pkgutil` and `kpt/internal/pkg` should probably be combined.

I want to get the behavior of get and update correct before trying to use the Package type. Otherwise we will end up spending time refactoring code that we plan to rewrite anyway.

File: internal/cmdcomplete/complete.go
Comment (mortent) [reply internal/cmdcomplete/complete.go:63]: I'm not sure. It is something we should make work when we have the new set of commands in place.

File: internal/util/fetch/fetch.go
Comment (mortent) [reply internal/util/fetch/fetch.go:37]: Plan on looking at this in a follow-up PR.

File: pkg/api/kptfile/v1alpha2/types.go
Comment (mortent) [reply pkg/api/kptfile/v1alpha2/types.go:43]: I think keeping information that is managed by kpt separate from the information we expect users to update makes it easier to understand.

File: internal/util/pkgutil/pkgutil.go
Comment (droot) [inline internal/util/pkgutil/pkgutil.go:33]: I think we have a method for returning direct packages with new pkg abstraction. See if you can use that ?

File: internal/cmdupdate/cmdupdate.go
Comment (frankfarzan) [reply internal/cmdupdate/cmdupdate.go:91]: Yes, computations should be based on `UniquePath`. `DisplayPath` is needed because different commands will display different information to users which is orthogonal to cobra or any choice of CLI framework for that matter. 

I'm not following why nested packages can't use package abstraction. Hydration logic already uses it in its recursive algorithm. But we can extend it with anything missing. It's a starting point for cleaning up the codebase, not exhaustive library.

I'm fine with deferring this in follow up with PRs to unblock things. I'm going to abstain from approving this and look at the follow up PRs instead.

File: internal/util/diff/diff.go
Comment (frankfarzan) [reply internal/util/diff/diff.go:307]: General note: having `fooutil` and `barutil` is usually a code smell. Aside from `util` being a meaningless term (Go naming practices advise against it), it usually signals that there we don't have the right abstractions in place.

**Review by droot — COMMENTED** [review]
I glanced through the code, looks reasonable to me on a high level.

File: internal/util/pkgutil/pkgutil.go
Comment (mortent) [reply internal/util/pkgutil/pkgutil.go:33]: Yeah, but there are some issues with using it. It is currently tied the package abstraction, which will require some larger changes to incorporate into the packaging code. It also isn't actually correct, since it only looks for subpackages in the immediate subdirectories. Trying to reconcile the different functions for traversing package hierarchies is on my list.

File: internal/cmdcomplete/complete.go
Comment (phanimarupaka) [reply internal/cmdcomplete/complete.go:63]: Yes. That would be the last step and treated as a nice to have.

File: internal/cmdupdate/cmdupdate.go
Comment (mortent) [reply internal/cmdupdate/cmdupdate.go:91]: I'm not saying it is not possible, it just requires some work. Packaging is different from hydration in we are building and updating the package hierarchy as we are traversing it so it needs a slightly different approach.

File: internal/util/get/get.go
Comment (mortent) [reply internal/util/get/get.go:75]: Yeah, the idea is to do this in "two phases". First update the desired state and then call the functionality for "reconciling" the actual state (in this case the package on disk) with the desired state. This supports use-cases where the desired state is updated by other tools or by users manually, or by kpt porcelain cli commands.

File: internal/testutil/setup_manager.go
Comment (mortent) [reply internal/testutil/setup_manager.go]: yeah, fixed this. Ref discussion further down, maybe we want to eventually find a shorter alias, but this makes things consistent.

File: internal/util/pkgutil/pkgutil.go
Comment (mortent) [reply internal/util/pkgutil/pkgutil.go]: This was unfortunate naming of the variables. I have updated the code and added some more comments.

File: internal/util/pkgutil/pkgutil.go
Comment (mortent) [reply internal/util/pkgutil/pkgutil.go]: Added a full description of this function.

File: pkg/kptfile/kptfileutil/util.go
Comment (mortent) [reply pkg/kptfile/kptfileutil/util.go:208]: For now, we only handle `upstream` and `upstreamLock` in the simplest way, so it is not yet needed. But when we add code for merging of the pipeline, we will definitely need it.

File: internal/util/diff/diff.go
Comment (mortent) [reply internal/util/diff/diff.go:307]: Agree there are too many `util` packages in kpt now. We should create more focused packages and remove the duplication.

File: internal/cmdupdate/cmdupdate.go
Comment (mortent) [reply internal/cmdupdate/cmdupdate.go:80]: Yeah, we should. I will do that in a follow-up

File: internal/cmdupdate/cmdupdate.go
Comment (mortent) [reply internal/cmdupdate/cmdupdate.go]: Good point. It makes sense that we should never use the `DisplayPath` for computations. I have updated the PR.

**Comment by phanimarupaka** [comment]
Tested this PR with no major issues. The Kptfile will have 2 sections upstream and upstreamLock, users should modify upstream section so that upstreamLock and package are synced with upstream. 

e.g.

```
apiVersion: kpt.dev/v1alpha2
kind: Kptfile
metadata:
  name: mysql
upstream:
  type: git
  git:
    repo: git@github.com:phanimarupaka/mysql
    directory: /
    ref: master
  updateStrategy: resource-merge
upstreamLock:
  type: git
  gitLock:
    repo: git@github.com:phanimarupaka/mysql
    directory: /
    ref: master
    commit: 435484eb7753aa00770cbe99152147775b158fe4
```

kpt hangs while fetching large repos. (https://github.com/cockroachdb/cockroach.git)

**Comment by phanimarupaka** [comment]
LGTM for me.

File: internal/cmdupdate/cmdupdate.go
Comment (frankfarzan) [inline internal/cmdupdate/cmdupdate.go:95]: Can you add a TODO to move the logic in 95-98 (including resolveRelPath) to Package.

File: internal/util/pkgutil/pkgutil.go
Comment (frankfarzan) [reply internal/util/pkgutil/pkgutil.go:33]: Can you add TODOs for future refactoring here

File: internal/cmdupdate/cmdupdate.go
Comment (frankfarzan) [inline internal/cmdupdate/cmdupdate.go:100]: In english, what is it that's being tested here?

Also, what do you expect the user input PKG_PATH to be? Can it be absolute or just relative? Where's the documentation for the semantic of this arg? @phanimarupaka 

If PKG_PATH is absolute: In general, when there are symlinks present (let's say in one of the sub-packages), we no longer have a hierarchical fs. So simply doing a `HasPrefix` on the UniquePath won't be correct.

If PKG_PATH is relative: then this check is easier.

This is the reason I strongly suggest against doing any path-related computation outside of a centralized, well-encapsulated library. It's subtle and error prone.

Comment (frankfarzan) [reply internal/cmdupdate/cmdupdate.go:100]: As mentioned before, not blocking this PR, but just assume this is buggy for now until we do it properly in future PRs.

Comment (mortent) [reply internal/cmdupdate/cmdupdate.go:102]: I think there are several questions here:

 * This functionality enforces that it is only possible to get/update package in subdirectories of cwd. We should decide if this check is needed.
   * If it is, then we should probably enforce it for all commands in a single place, so probably in the top-level kpt cobra command.
 * The provided path can be either absolute or relative, but we want the internals of kpt to work with absolute paths. We should handle this conversion as early as possible and have a single library for handling it.
 * Appropriate handling of symlinks is something we need to look at for all functionality in kpt. This probably includes understanding how it is handled in git.

File: internal/cmdupdate/cmdupdate.go
Comment (mortent) [reply internal/cmdupdate/cmdupdate.go:97]: done

File: internal/util/pkgutil/pkgutil.go
Comment (mortent) [reply internal/util/pkgutil/pkgutil.go:36]: Done. Some of this work is already in https://github.com/GoogleContainerTools/kpt/pull/1583

**Review by phanimarupaka — APPROVED** [review]

**Pull request merged**
